### ISSUE TITLE
Import hangs in combination with PyTorch dependencies torchtext/torchvision

### ISSUE THREAD

**Issue by schoonhovenrichard** [scverse/anndata#1679]
### Please make sure these conditions are met

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the latest version of anndata.
- [x] (optional) I have confirmed this bug exists on the master branch of anndata.

### Report

**Behaviour:**

```python
import anndata
```

hangs indefintely.

**Steps to reproduce**
```bash
conda create -n test pytorch=2.4.1 pytorch-cuda torchtext==0.6.0 torchvision=0.19.1 pip -c nvidia -c pytorch -c defaults -c conda-forge
conda activate test
pip install anndata==0.10.9
```
then
```python
import anndata
```
hangs indefinitely.

**Fix:**
Simple fix is 
```bash
pip install anndata=0.8.0
```

**Environment:**
Using conda environment that contains:

```bash
pytorch                   2.4.1           py3.12_cuda12.4_cudnn9.1.0_0    pytorch
pytorch-cuda              12.4                 hc786d27_6    pytorch
pytorch-mutex             1.0                        cuda    pytorch
torchtext                 0.6.0                      py_1    pytorch
torchtriton               3.0.0                     py312    pytorch
torchvision               0.19.1              py312_cu124    pytorch
....
anndata                   0.10.9                   pypi_0    pypi
```

**Verbose trace**
Maybe a hint, the trace seems to come from anndata.experimental.pytorch
```bash
python -v -c "import anndata"
```

gives the following trace:

```bash
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/anndata/experimental/pytorch/__pycache__/_annloader.cpython-312.pyc'
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/__pycache__/__init__.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/__init__.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/__pycache__/__init__.cpython-312.pyc'
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/__pycache__/pdb.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/pdb.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/__pycache__/pdb.cpython-312.pyc'
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/__pycache__/cmd.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/cmd.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/__pycache__/cmd.cpython-312.pyc'
import 'cmd' # <_frozen_importlib_external.SourceFileLoader object at 0x7f55220ca660>
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/__pycache__/bdb.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/bdb.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/__pycache__/bdb.cpython-312.pyc'
import 'bdb' # <_frozen_importlib_external.SourceFileLoader object at 0x7f55220cb0b0>
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/__pycache__/code.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/code.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/__pycache__/code.cpython-312.pyc'
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/__pycache__/codeop.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/codeop.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/__pycache__/codeop.cpython-312.pyc'
import 'codeop' # <_frozen_importlib_external.SourceFileLoader object at 0x7f55220fcec0>
import 'code' # <_frozen_importlib_external.SourceFileLoader object at 0x7f55220fc680>
import 'pdb' # <_frozen_importlib_external.SourceFileLoader object at 0x7f55220b74d0>
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/__pycache__/_utils.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_utils.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/__pycache__/_utils.cpython-312.pyc'
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/__pycache__/typing_extensions.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/typing_extensions.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/__pycache__/typing_extensions.cpython-312.pyc'
import 'typing_extensions' # <_frozen_importlib_external.SourceFileLoader object at 0x7f55220ff680>
import 'torch._utils' # <_frozen_importlib_external.SourceFileLoader object at 0x7f55220fdf10>
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/__pycache__/_utils_internal.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_utils_internal.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/__pycache__/_utils_internal.cpython-312.pyc'
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_strobelight/__pycache__/__init__.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_strobelight/__init__.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_strobelight/__pycache__/__init__.cpython-312.pyc'
import 'torch._strobelight' # <_frozen_importlib_external.SourceFileLoader object at 0x7f552211d340>
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_strobelight/__pycache__/compile_time_profiler.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_strobelight/compile_time_profiler.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_strobelight/__pycache__/compile_time_profiler.cpython-312.pyc'
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_strobelight/__pycache__/cli_function_profiler.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_strobelight/cli_function_profiler.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_strobelight/__pycache__/cli_function_profiler.cpython-312.pyc'
import 'torch._strobelight.cli_function_profiler' # <_frozen_importlib_external.SourceFileLoader object at 0x7f552211d5e0>
import 'torch._strobelight.compile_time_profiler' # <_frozen_importlib_external.SourceFileLoader object at 0x7f552211c9e0>
import 'torch._utils_internal' # <_frozen_importlib_external.SourceFileLoader object at 0x7f55220ffc50>
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/__pycache__/torch_version.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/torch_version.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/__pycache__/torch_version.cpython-312.pyc'
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_vendor/__pycache__/__init__.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_vendor/__init__.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_vendor/__pycache__/__init__.cpython-312.pyc'
import 'torch._vendor' # <_frozen_importlib_external.SourceFileLoader object at 0x7f552211ed50>
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_vendor/packaging/__pycache__/__init__.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_vendor/packaging/__init__.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_vendor/packaging/__pycache__/__init__.cpython-312.pyc'
import 'torch._vendor.packaging' # <_frozen_importlib_external.SourceFileLoader object at 0x7f552211eb10>
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_vendor/packaging/__pycache__/version.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_vendor/packaging/version.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_vendor/packaging/__pycache__/version.cpython-312.pyc'
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_vendor/packaging/__pycache__/_structures.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_vendor/packaging/_structures.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/_vendor/packaging/__pycache__/_structures.cpython-312.pyc'
import 'torch._vendor.packaging._structures' # <_frozen_importlib_external.SourceFileLoader object at 0x7f5522160200>
import 'torch._vendor.packaging.version' # <_frozen_importlib_external.SourceFileLoader object at 0x7f552211ee70>
# /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/__pycache__/version.cpython-312.pyc matches /home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/version.py
# code object from '/home/schoonho/anaconda3/envs/transformerGPU/lib/python3.12/site-packages/torch/__pycache__/version.cpython-312.pyc'
import 'torch.version' # <_frozen_importlib_external.SourceFileLoader object at 0x7f5522160da0>
import 'torch.torch_version' # <_frozen_importlib_external.SourceFileLoader object at 0x7f552211e2a0>
....(here it hangs indefinitely)
```

### Versions

```python
So this is not possible due to the fact that the import itself fails unfortunately.
```

**Comment by ilan-gold** [comment]
That is strange.  @Intron7 Do you have any thoughts on this at first glance?  Something to do with GPU that you've seen maybe?

**Comment by Intron7** [comment]
I can reproduce your error to a certain extend. However when I try to load this in python/ jupyter it works like a charm with `anndata 0.10.9`. The CLI run still breaks after downgrading to `anndata 0.8.0` at least for me. I don't know but I looks like its some dependency stuff.

**Comment by Intron7** [comment]
Can you try to install anndata with the other conda packages we have a working conda-forge package. There is works for me also in CLI

**Comment by schoonhovenrichard** [comment]
Thanks for looking into this. I am now on a different system and I tested this again and I can reproduce this problem. Also a colleague had a similar problem I believe.

This works:
```bash
conda create -n test pytorch=2.4.1 pytorch-cuda pip -c nvidia -c pytorch -c defaults -c conda-forge
conda activate test
pip install anndata==0.10.9
```

and then import anndata in python.

But this does not:

```bash
conda create -n test pytorch=2.4.1 pytorch-cuda torchtext==0.6.0 torchvision=0.19.1 pip -c nvidia -c pytorch -c defaults -c conda-forge
conda activate test
pip install anndata==0.10.9
```

So it is the inclusion of torchtext or torchvision that is causing the problem.

**Comment by schoonhovenrichard** [comment]
For an exact copy of the environment:
```bash
name: test
channels:
  - pytorch
  - nvidia
  - defaults
dependencies:
  - _libgcc_mutex=0.1=main
  - _openmp_mutex=5.1=1_gnu
  - blas=1.0=mkl
  - brotli-python=1.0.9=py312h6a678d5_8
  - bzip2=1.0.8=h5eee18b_6
  - ca-certificates=2024.7.2=h06a4308_0
  - certifi=2024.8.30=py312h06a4308_0
  - charset-normalizer=3.3.2=pyhd3eb1b0_0
  - cuda-cudart=12.4.127=0
  - cuda-cupti=12.4.127=0
  - cuda-libraries=12.4.0=0
  - cuda-nvrtc=12.4.127=0
  - cuda-nvtx=12.4.127=0
  - cuda-opencl=12.6.68=0
  - cuda-runtime=12.4.0=0
  - cuda-version=12.6=3
  - expat=2.6.3=h6a678d5_0
  - ffmpeg=4.3=hf484d3e_0
  - filelock=3.13.1=py312h06a4308_0
  - freetype=2.12.1=h4a9f257_0
  - gmp=6.2.1=h295c915_3
  - gnutls=3.6.15=he1e5248_0
  - idna=3.7=py312h06a4308_0
  - intel-openmp=2023.1.0=hdb19cb5_46306
  - jinja2=3.1.4=py312h06a4308_0
  - jpeg=9e=h5eee18b_3
  - lame=3.100=h7b6447c_0
  - lcms2=2.12=h3be6417_0
  - ld_impl_linux-64=2.38=h1181459_1
  - lerc=3.0=h295c915_0
  - libcublas=12.4.2.65=0
  - libcufft=11.2.0.44=0
  - libcufile=1.11.1.6=0
  - libcurand=10.3.7.68=0
  - libcusolver=11.6.0.99=0
  - libcusparse=12.3.0.142=0
  - libdeflate=1.17=h5eee18b_1
  - libffi=3.4.4=h6a678d5_1
  - libgcc-ng=11.2.0=h1234567_1
  - libgomp=11.2.0=h1234567_1
  - libiconv=1.16=h5eee18b_3
  - libidn2=2.3.4=h5eee18b_0
  - libjpeg-turbo=2.0.0=h9bf148f_0
  - libnpp=12.2.5.2=0
  - libnvfatbin=12.6.68=0
  - libnvjitlink=12.4.99=0
  - libnvjpeg=12.3.1.89=0
  - libpng=1.6.39=h5eee18b_0
  - libstdcxx-ng=11.2.0=h1234567_1
  - libtasn1=4.19.0=h5eee18b_0
  - libtiff=4.5.1=h6a678d5_0
  - libunistring=0.9.10=h27cfd23_0
  - libuuid=1.41.5=h5eee18b_0
  - libwebp-base=1.3.2=h5eee18b_0
  - llvm-openmp=14.0.6=h9e868ea_0
  - lz4-c=1.9.4=h6a678d5_1
  - markupsafe=2.1.3=py312h5eee18b_0
  - mkl=2023.1.0=h213fc3f_46344
  - mkl-service=2.4.0=py312h5eee18b_1
  - mkl_fft=1.3.10=py312h5eee18b_0
  - mkl_random=1.2.7=py312h526ad5a_0
  - mpmath=1.3.0=py312h06a4308_0
  - ncurses=6.4=h6a678d5_0
  - nettle=3.7.3=hbbd107a_1
  - networkx=3.3=py312h06a4308_0
  - numpy=1.26.4=py312hc5e2394_0
  - numpy-base=1.26.4=py312h0da6c21_0
  - openh264=2.1.1=h4ff587b_0
  - openjpeg=2.5.2=he7f1fd0_0
  - openssl=3.0.15=h5eee18b_0
  - pillow=10.4.0=py312h5eee18b_0
  - pip=24.2=py312h06a4308_0
  - pysocks=1.7.1=py312h06a4308_0
  - python=3.12.5=h5148396_1
  - pytorch=2.4.1=py3.12_cuda12.4_cudnn9.1.0_0
  - pytorch-cuda=12.4=hc786d27_6
  - pytorch-mutex=1.0=cuda
  - pyyaml=6.0.1=py312h5eee18b_0
  - readline=8.2=h5eee18b_0
  - requests=2.32.3=py312h06a4308_0
  - setuptools=72.1.0=py312h06a4308_0
  - six=1.16.0=pyhd3eb1b0_1
  - sqlite=3.45.3=h5eee18b_0
  - sympy=1.13.2=py312h06a4308_0
  - tbb=2021.8.0=hdb19cb5_0
  - tk=8.6.14=h39e8969_0
  - torchtext=0.6.0=py_1
  - torchtriton=3.0.0=py312
  - torchvision=0.19.1=py312_cu124
  - tqdm=4.66.5=py312he106c6f_0
  - typing_extensions=4.11.0=py312h06a4308_0
  - urllib3=2.2.2=py312h06a4308_0
  - wheel=0.44.0=py312h06a4308_0
  - xz=5.4.6=h5eee18b_1
  - yaml=0.2.5=h7b6447c_0
  - zlib=1.2.13=h5eee18b_1
  - zstd=1.5.5=hc292b87_2
  - pip:
      - anndata==0.10.9
      - array-api-compat==1.8
      - h5py==3.11.0
      - natsort==8.4.0
      - packaging==24.1
      - pandas==2.2.2
      - python-dateutil==2.9.0.post0
      - pytz==2024.2
      - scipy==1.14.1
      - tzdata==2024.1
prefix: /home/schoonho/anaconda3/envs/test
```

**Comment by schoonhovenrichard** [comment]
> Can you try to install anndata with the other conda packages we have a working conda-forge package. There is works for me also in CLI

@Intron7 If I install the conda package the same problem occurs.

```bash
conda create -n test pytorch=2.4.1 pytorch-cuda torchtext==0.6.0 torchvision=0.19.1 pip anndata -c nvidia -c pytorch -c defaults -c conda-forge
```

**Comment by Intron7** [comment]
@schoonhovenrichard which os are you using. I use ubuntu-22.04 and I dont get the error.

**Comment by schoonhovenrichard** [comment]
@Intron7 I am also on Ubuntu 22.04.4 LTS, Nvidia Driver Version: 555.42.06, (CUDA Version: 12.5 but it is irrelevant as I have a new CUDA inside the conda env). So if you copy my exact environment into a .yml file and create a conda environment from that, and then try import anndata it does not hang?

**Comment by Intron7** [comment]
@schoonhovenrichard I'm on Ubuntu 22.04.5 Driver Version: 560.35.03 CUDA Version: 12.6. And your exact environment works for me.

**Comment by schoonhovenrichard** [comment]
@Intron7 Okay then it is very strange. Maybe we should close this issue for now? It's strange that on two different systems I get this error but again, if its not reproducible we can close this for now. If someone else runs into the same problem we could re-open it?

**Comment by Intron7** [comment]
@ilan-gold what do you think shall we close this?

**Comment by ilan-gold** [comment]
Not sure - https://github.com/scverse/scvi-tools/blob/e8f9b369865e80567d5e900d32d8e7731497b974/Dockerfile If you have access to docker @schoonhovenrichard maybe see if the same things happens with scvi-tools?  If not, then yes, I think it's tough to do much else.

**Comment by ilan-gold** [comment]
I will also try this env out myself once quickly.

**Comment by ilan-gold** [comment]
Separately, @schoonhovenrichard you could try these commands with the anndata deps individually and see if it's one of them causing it - there aren't many, and some are likely shared with `torch`.  Short of that, not sure there's much else.  I also can't reproduce this given what's here.

**Comment by schoonhovenrichard** [comment]
@Intron7 @ilan-gold So I think it is unrelated to anndata mostly. I opened a PyTorch post here:
https://discuss.pytorch.org/t/broken-torchtext-conda-package/209794

which is purely combining pytorch with torchtext. Later I noticed that sometimes conda pulls the conda-forge pytorch and not pytorch channel version. I have not tested this, but maybe thorugh scvi-tools or something a similar thing happens in other cases. Either way, the hanging behaviour is potentially explained by the response, i.e., CUDA JIT trying to compile it on the fly.
